src/python/pants/engine/internals/synthetic_targets.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+"""Synthetic targets is a concept of injecting targets into the build graph that doesn't have a home
+in any BUILD file.
+
+This is achieved by declaring a union member for the `SyntheticTargetAdaptorsRequest` union along
+with a rule taking that member and returning a `SyntheticTargetAdaptors`.
+
+Example demonstarting how to register synthetic targets:
+
+    from pants.engine.internals.synthetic_targets import (
+        SyntheticTargetAdaptor,
+        SyntheticTargetAdaptors,
+        SyntheticTargetAdaptorsRequest,
+    )
+    from pants.engine.unions import UnionRule
+    from pants.engine.rules import collect_rules, rule
+
+
+    class SyntheticExampleTargetAdaptorsRequest(SyntheticTargetAdaptorsRequest):
+        pass
+
+
+    @rule
+    async def example_synthetic_targets(_: SyntheticExampleTargetAdaptorsRequest) -> SyntheticTargetAdaptors:
```

Comment:
Even though we will almost always end up generate all targets in the repository, I think that I would still feel better about this API if it generated targets directory-at-a-time.

I also think that that could potentially be easier to integrate with the `AddressMap` parsing, since you could essentially assert that the generator only generated targets living in the directory, and then stitch them into the `AddressMap`?

Reply:
> Even though we will almost always end up generate all targets in the repository, I think that I would still feel better about this API if it generated targets directory-at-a-time.

Oh: I forgot the obvious case where this isn't true: during Specs expansion.

So yea, I think that it is worthwhile.

Reply:
Reworked into requesting synth target adaptors on a per directory basis.

Reply:
Reverted back again, sorry for the back and forth.
But I can't make this into an API that I find appealing when I get a request per path.

Consider this example, using a single request for the entire repo:
```python
@rule
async def python_lockfile_synthetic_targets(
    _: PythonSyntheticLockfilesTargetAdaptorsRequest,
    python_setup: PythonSetup,
) -> SyntheticTargetAdaptors:
    return SyntheticTargetAdaptors(
        [
            SyntheticTargetAdaptor.create(
                path=os.path.dirname(lockfile),
                type_alias="lockfiles",
                name=name,
                sources=[os.path.basename(lockfile)],
            )
            for name, lockfile in python_setup.resolves.items()
        ]
        if python_setup.enable_resolves
        else []
    )

```

That feels so much more efficient, than if I were to get called once for every directory with a BUILD file in it..

Reply:
I think that the first use case that you've chosen here has a particular shape, where all of the targets are known globally ahead of time.

Any use case which was creating targets for files (say: automatically generating targets for Python files) which actually needed discovery would need to start by scanning the entire repository with a glob like `PathGlobs(['**/*.py'])`. Whereas with a directory-centric API they could scan only the relevant directory.

Reply:
Hmm... interesting. But then there would need to be a BUILD file in every directory where you want to look for synthetic targets (or use some other mechanism for triggering the collection of synthetic targets) but maybe that's fine.

In my mind, the case for synthetic targets is more about knowing before hand what targets there is you want to create, and less about scanning for them, but that may of course not be true. I'm thinking that these two cases may be a candidate to have as a class var then on the union member, to choose which method to apply that will be more efficient.